### PR TITLE
Java References tasks

### PR DESCRIPTION
**Pull request by NataliaBodrug**
Java References tasks: (work with Arrays, DateTime, Regex)

### CODE REVIEW

File: src/test/java/javaReferences/Task1_ReverseArray.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import java.util.Scanner;
+
+/**
+ * Task1: Avem un array unidimensional(a[i]). Copiem valorile din array-ul dat in altul in ordine inversata.
+ */
+
+public class Task1_ReverseArray {
+    public static void main(String[] args) {
+        reverseArray();
+        reverseArrayUsingScanner();
```
Comment (am-bee73) [inline src/test/java/javaReferences/Task1_ReverseArray.java:12]: with or without Scanner it's quite same, please try to keep code without duplicates

Comment (NataliaBodrug) [reply src/test/java/javaReferences/Task1_ReverseArray.java:12]: ok, will remove first method.

Comment (NataliaBodrug) [reply src/test/java/javaReferences/Task1_ReverseArray.java:11]: done.

File: src/test/java/javaReferences/Task1_ReverseArray.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+        int arr1[] = new int[len];
+        System.out.println("Please insert " + len + " values: ");
+
+        for (int i = 0; i < len; i++) {
+            arr1[i] = sc.nextInt();
+        }
+
+        System.out.println("Print initial array:");
+        for (int i : arr1) {
```
Comment (FroggyFeet) [inline src/test/java/javaReferences/Task1_ReverseArray.java:28]: Take a look at Arrays.toString(..) for printing arrays

File: src/test/java/javaReferences/Task1_ReverseArray.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        int arr2[] = new int[len];
+
+        for (int i = 0; i < arr1.length; i++) {
+            arr2[i] = arr1[len - i - 1];
+        }
+
+        System.out.println("\nPrint reversed array: ");
+        for (int i : arr2) {
```
Comment (FroggyFeet) [inline src/test/java/javaReferences/Task1_ReverseArray.java:39]: The same as above

File: src/test/java/javaReferences/Task3_SumOfRowsAndColumnsFromArray.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            for (int j = 0; j < array.length; j++) {
+                System.out.print(array[i][j] + " ");
+            }
+            System.out.println();
+        }
+
+        System.out.println("Print the sum of each row:");
+        int sumRow = 0;
+
+        for (int i = 0; i < array.length; i++) {
```
Comment (FroggyFeet) [inline src/test/java/javaReferences/Task3_SumOfRowsAndColumnsFromArray.java:29]: Let's extract this logic to a separate method. We will have a method which calculates sum of rows, and another one for columns.

File: src/test/java/javaReferences/Task4_Swap_UpperLowerCases.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+    static void swapCharacterUpperLowerCases() {
+        String text = "Natalia Is Learning Java";
+        System.out.println("The initial text is: " + text);
+
+        String newText = "";
+        for (int i = 0; i < text.length(); i++) {
+            if (Character.isLowerCase(text.charAt(i))) {
+                newText = newText + Character.toUpperCase(text.charAt(i));
+            } else if (Character.isUpperCase(text.charAt(i))) {
```
Comment (FroggyFeet) [inline src/test/java/javaReferences/Task4_Swap_UpperLowerCases.java:23]: We don't need to validate it twice. If it is not a lower case character, than we should make it lower case, and that's it :)

File: src/test/java/javaReferences/Task5_LastAndNextFriday.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+    public static void main(String[] args) {
+
+        printLastAndNextFriday();
+    }
+
+    static void printLastAndNextFriday() {
+
+        LocalDate dt = LocalDate.now();
+        System.out.println("Next Friday: " + dt.with(TemporalAdjusters.next(DayOfWeek.FRIDAY)));
```
Comment (FroggyFeet) [inline src/test/java/javaReferences/Task5_LastAndNextFriday.java:21]: Let's use String.format(..) for parameterized strings.

File: src/test/java/javaReferences/Task7_DateTimeFormats.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    public static void main(String[] args) {
+
+        printDateTimeFormats();
+    }
+
+    static void printDateTimeFormats() {
+
+        Date date = new Date();
+        System.out.println("Print format: DD::MMM::YYYY  e.g.(16::JUN::2020):\n");
+        SimpleDateFormat formatter1 = new SimpleDateFormat("d::MMM::yyyy");
```
Comment (FroggyFeet) [inline src/test/java/javaReferences/Task7_DateTimeFormats.java:27]: It's Ok to use SimpleDateFormat. However, DateTimeFormatter is used more often. Try to change the implementation in order to use use it.

Comment (FroggyFeet) [reply src/test/java/javaReferences/Task7_DateTimeFormats.java:27]: And let's extract printing into a separate method. We will have a method (something like "printFormattedDate(LocalDateTime dateTime, String format)").
And all the magic will happen there

File: src/test/java/javaReferences/Task8_Regex.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+            if (text2[i].matches("(.)*[aA](.)*") && text2[i].matches("(.)*[oO](.)*") && text2[i].matches("(.)*[eE](.)*")) {
+                lettersAEO += text2[i] + "\n";
+            }
+        }
+        System.out.println(lettersAEO);
+
+        System.out.println("\nWords that contain more than 6 letters: \n");
+        Pattern sevenLetters = Pattern.compile("\\S[a-zA-Z0-9]{6,}"); // or \S[a-zA-Z]{6,}  or \w{6,}
+        Matcher matchSevenLetters = sevenLetters.matcher(text);
```
Comment (FroggyFeet) [inline src/test/java/javaReferences/Task8_Regex.java:49]: Let's name variable more clear) In fact "Matcher matchSevenLetter" is "Matcher moreThanSixLettersMatcher". As you don't look for 7-letter-words, but for words that contain more than 6 letters.

File: src/test/java/javaReferences/Task4_Swap_UpperLowerCases.java (hunk 5 above)
Comment (NataliaBodrug) [reply src/test/java/javaReferences/Task4_Swap_UpperLowerCases.java:23]: hm, i don't think so... 
In requirements we have: 
"Task4: Avem o variabila de tip string, careia ii atribuim o fraza. Schimbam literele mari cu acelea mici, si invers." 

This word "invers" means that if we have lower case - change to upper, if we have upper case-change to lower. 

Please correct me, if I understood wrong

File: src/test/java/javaReferences/Task8_Regex.java (hunk 8 above)
Comment (NataliaBodrug) [reply src/test/java/javaReferences/Task8_Regex.java:49]: done.

File: src/test/java/javaReferences/Task1_ReverseArray.java (hunk 2 above)
Comment (NataliaBodrug) [reply src/test/java/javaReferences/Task1_ReverseArray.java:28]: changed.

File: src/test/java/javaReferences/Task1_ReverseArray.java (hunk 3 above)
Comment (NataliaBodrug) [reply src/test/java/javaReferences/Task1_ReverseArray.java:39]: will keep this time (just to have foreach method too). 
For printing initial array, changed in sout with Arrays.toString(arr1).

File: src/test/java/javaReferences/Task3_SumOfRowsAndColumnsFromArray.java (hunk 4 above)
Comment (NataliaBodrug) [reply src/test/java/javaReferences/Task3_SumOfRowsAndColumnsFromArray.java:29]: created methods: 
sumOfEachRow(array);
sumOfEachColumn(array);

File: src/test/java/javaReferences/Task4_Swap_UpperLowerCases.java (hunk 5 above)
Comment (NataliaBodrug) [reply src/test/java/javaReferences/Task4_Swap_UpperLowerCases.java:23]: done some refactoring in all methods.

File: src/test/java/javaReferences/Task5_LastAndNextFriday.java (hunk 6 above)
Comment (NataliaBodrug) [reply src/test/java/javaReferences/Task5_LastAndNextFriday.java:21]: added method: printLastAndNextFridayUsingStringFormat(LocalDate dt);
done some refactoring on both methods;

File: src/test/java/javaReferences/Task7_DateTimeFormats.java (hunk 7 above)
Comment (NataliaBodrug) [reply src/test/java/javaReferences/Task7_DateTimeFormats.java:27]: Added method: printFormattedDate(LocalDateTime localDateTime, String format) for DateTimeFormatter implementation.

Some refactoring on initial  printDateTimeFormats() method;

Keep only one example with  SimpleDateFormat(old version) in new method printThroughSimpleDateFormatter().

Comment (NataliaBodrug) [reply src/test/java/javaReferences/Task7_DateTimeFormats.java:27]: PS: Magic sounds perfect. 💯

**Comment by NataliaBodrug** [comment]
Refactoring on task8_Regex

**Comment by NataliaBodrug** [comment]
> Refactoring on Task8_Regex

**Comment by NataliaBodrug** [comment]
Refactoring on Task8_regex; split in methods;

**Comment by FroggyFeet** [comment]
Don't you want to have 1 commit instead of 9?)

**Comment by NataliaBodrug** [comment]
> Don't you want to have 1 commit instead of 9?)

I tried with multiple commits in special ( just wanted to see how they will look like in Git after merge),
is this bad practice?

**Comment by NataliaBodrug** [comment]
> > Don't you want to have 1 commit instead of 9?)
> 
> I tried with multiple commits in special ( just wanted to see how they will look like in Git after merge),
> is this bad practice?

-looks ugly in Git logs :)  -> changed to less commits;

**Pull request merged**